### PR TITLE
Supported using custom map implementation for JSONObject

### PR DESCRIPTION
**Pull request by lvca**
Supported using custom map implementation for JSONObject, like a LinkedHashMap to keep the insertion order to the map fields.

This is a more flexible implementation than https://github.com/stleary/JSON-java/pull/655.

### CODE REVIEW

**Comment by lvca** [comment]
The rationale behind using this approach was:
- I like to keep `map` property final for a lot of reasons, and replacing the internal map after the default instantiation of the HashMap was overkill
- I couldn't use `JSONObject(Map)` constructor because it's already used to copy a map into the object
- Having a protected constructor mandate extending the JSONObject class
- This allows also using hyper optimized map implementation (like Trove collections or Guava) if needed

**Comment by johnjaylward** [comment]
I think this is a good start, but it still doesn't have a solution for JSONTokener to use either a custom map property itself, or the new optional extended class like you have in your test case.

Would you be willing to look into what changes would be needed to extend JSONTokener to optionally use the extended JSONObject (i.e. `OrderedJSONObject`) as part of some kind of configuration on the Tokener?

Basically this line would be the one needing to change based on how you set up the configuration: https://github.com/stleary/JSON-java/blob/master/src/main/java/org/json/JSONTokener.java#L431

**Comment by stleary** [comment]
It might be better to get a consensus on the design before submitting a pull request. Issue #571 has been kept open and is available for this purpose.

**Comment by lvca** [comment]
Updated the PR with the support for JSONTokener and wrote on https://github.com/stleary/JSON-java/issues/571.

File: src/main/java/org/json/JSONObject.java [hunk 1]
```diff
@@ -208,16 +208,16 @@ public JSONObject(JSONObject jo, String ... names) {
...
+         *
+         * @param x
+         *            A JSONTokener object containing the source string.
+         * @throws JSONException
+         *             If there is a syntax error in the source string or a
+         *             duplicated key.
+         */
     public JSONObject(JSONTokener x) throws JSONException {
-        this();
+        this(x.getMapImplementation());
```
Comment (abimarank) [inline src/main/java/org/json/JSONObject.java:220]: Better do a null check before invoking on that instance

**Review by abimarank — CHANGES_REQUESTED** [review]

File: src/main/java/org/json/JSONObject.java (hunk 1 above)
Comment (johnjaylward) [reply src/main/java/org/json/JSONObject.java:220]: The tokener is expected to be non-null always there, the the called constructor checks for a null parameter already. This looks fine as-is, but possibly ensure the java doc calls out the expectation.

**Review by johnjaylward — COMMENTED** [review]
Overall, this looks like a fairly simple change that matches my expectations. I will need more time to review more thoroughly to ensure we have consistent API across the classes due to the holidays.

@stleary can we leave this up for review through mid January?

**Comment by stleary** [comment]
This code will remain in review for a minimum of 3 weeks before a decision is made.

**Comment by lvca** [comment]
Any news on this? Just checking.

**Review by johnjaylward — APPROVED** [review]
I think this looks good.

**Comment by aaron-613** [comment]
Are we finally getting ordered keys?  Nice!

**Comment by atkawa7** [comment]
@stleary Thank you for the response. I really wanted to use this for a special case. I was trying build a custom jsonschema2pojo generator. The issue when I am coming across when using this lib to parse the json is order of generated fields. I am expecting the order of fields to consistent with the order of jsonschema properties. However this is not the case. The order changes frequently. The generated code is not ignored when pushed to git. So frequent code generations result in different code arrangements which result in large diffs. 

```
{
  "$id": "https://example.com/person.schema.json",
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "title": "Person",
  "type": "object",
  "properties": {
    "firstName": {
      "type": "string",
      "description": "The person's first name."
    },
    "lastName": {
      "type": "string",
      "description": "The person's last name."
    },
    "age": {
      "description": "Age in years which must be equal to or greater than zero.",
      "type": "integer",
      "minimum": 0
    }
  }
}
```

```java
public interface Person{
  Integer age();
  String firstName();
  String lastName();
}
```

**Comment by johnjaylward** [comment]
@stleary , while I'm not saying this is the best implementation, it does provide a few benefits:

1. It's 100% backwards compatible with existing code seeing no change
2. We don't actually have an "ordered" implementation in this library.
3. It allows downstream users to create their own sub-classes that has their own order rules. i.e. some may want insertion order, others may want alphabetical.

The one thing I'm not in love of is the new public getter/setter on the JSONTokener (`getMapImplementation` and `setMapImplementation`). We really don't have an API similar to that used elsewhere in the library. I think I would prefer a constructor override there instead and make the value `final`.

What I'd also say we need to do with this PR if it's excepted is update the [Examples](https://github.com/stleary/JSON-java/blob/master/Examples.md) documentation to show a simple implementation.

**Comment by zjns** [comment]
In my opinion, it's not best solution, `JSONObject(int initialCapacity)` and `JSONObject(Map<?, ?> m)` constructor still can't be customized.

**Comment by stleary** [comment]
@zjns Any suggestions on how to improve this?

**Comment by zjns** [comment]
@stleary I'd prefer some codes like this.

MapFactory
```java
public interface MapFactory {
    MapFactory DEFAULT = new MapFactory() {
        @Override
        public Map<String, Object> newMap(int initialCapacity) {
            if (initialCapacity == -1) {
                return new HashMap<>();
            }
            return new HashMap<>(initialCapacity);
        }
    };

    Map<String, Object> newMap(int initialCapacity);
}
```

In JSONObject
```java
private MapFactory mapFactory = MapFactory.DEFAULT;

public JSONObject() {
        this.map = mapFactory.newMap(-1);
}

public JSONObject(Map<?, ?> m) {
    if (m == null) {
        this.map = mapFactory.newMap(-1);
    } else {
        this.map = mapFactory.newMap(m.size());
        # other codes
    }
}

protected JSONObject(int initialCapacity){
    this.map = mapFactory.newMap(initialCapacity);
}

public void setMapFactory(MapFactory mapFactory){
    this.mapFactory = mapFactory;
}

public MapFactory getMapFactory(){
    return mapFactory;
}
```

**Comment by stleary** [comment]
Does anyone have an opinion about whether something like @zjns' proposed implementation is better or worse than the original?

**Comment by lvca** [comment]
Guys, there is zero commitment to the project and zero support for any effort coming from the community. I wrote this PR almost 6 months ago, and nobody knows what to do with it. I got your feedback, changed the code, and sent it back. Still nothing. Who is in charge? Please, this is painful...

**Comment by stleary** [comment]
@lvca Sorry, my fault for letting this linger so long. The fact that this PR is still open means I haven't given up on it. 

An alternative map implementation goes against the design by the application author, and changes to preserve ordering have been rejected by him in the past.  This raises a high bar for a change to be accepted now.

I am willing to consider an update that meets these criteria:
- Does not have any impact on existing users
- Easy to use
- Generally useful for a range of different users
- Well designed, documented, written, and tested.

**Comment by javadev** [comment]
It will be better to replace HashMap with LinkedHashMap.

**Comment by stleary** [comment]
@javadev as a senior engineer, can you please explain your reasoning for recommending a LinkedHashMap, given that this is a reference application for parsing JSON in a way that follows the spec?

**Comment by javadev** [comment]
> @javadev as a senior engineer, can you please explain your reasoning for recommending a LinkedHashMap, given that this is a reference application for parsing JSON in a way that follows the spec?

I know about spec. It may be optional, I agree.

**Comment by stleary** [comment]
Please continue this discussion in #681. As soon as we have a PR that meets the requirements, it will be approved and merged. Apologies for taking so long to come to the decision to break with the direction set by the original author, who is also the person who wrote the JSON spec.

The requirements will be that the app must use the existing HashMap by default but can be configured or executed to run in a mode that uses a LinkedHashMap instead. The new mode should only use LinkedHashMap, and never use HashMap. I will come up with a suitable test suite for validation.

I am closing this PR because I don't think it complies. If you think differently, post here, and I will reopen PR and run the test suite against it (as soon as it's written). Otherwise, I think the approach outlined by @zjns may be a good place to start.

**Comment by lvca** [comment]
Hi guys, it's been more than 1y... @stleary I don't understand why my solution wasn't good enough. Could you please elaborate on it? You said you'd prefer a better solution, but how?
